### PR TITLE
fix: page router validations and add test cases

### PR DESCRIPTION
**Pull request by karrui**
This PR starts the chain of PRs for adding test cases for all our procedures in the Studio application so we can be more confident when shipping code.

## Solution

**Breaking Changes**
- [x] No - this PR is backwards compatible

**Features**:

- Details ...

**Improvements**:

- Add Tagged types to `jsonb` columns so we remember to call `jsonb()` function before insertion into the database. This prevents runtime errors.
- Update `updatePageBlobSchema` to validate the type of `content` in the schema itself.
- Update all `*select` constants to use `satisfies` keyword for better type inference when used.
- Throw 404 errors when resource is not found in all `page.router` procedures
- Validate whether `folderId` is a folder when calling `createPage` procedure and passing the `folderId` arg.
- use transaction when updating page in `updatePage` procedure
- add foreign key violation error handling

**Bug Fixes**:

- use correct `fullPage.publishedVersionId` check in `readPageAndBlob` procedure

## Tests

Tests are added for all procedures in the page router.

A few tests regarding permissions are skipped until permissions are in cc @seaerchin

### CODE REVIEW

**Comment by karrui** [comment]
* **#720** <a href="https://app.graphite.dev/github/pr/opengovsg/isomer/720?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @karrui and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: apps/studio/src/schemas/page.ts
Comment (harishv7) [inline apps/studio/src/schemas/page.ts:63]: we are trying to abort early right, do we need the FATAL: true flag?

File: apps/studio/src/server/modules/site/site.service.ts
Comment (harishv7) [inline apps/studio/src/server/modules/site/site.service.ts:41]: previously when we didn't cast it as jsonb it still worked right? how does the cast help now?

File: tooling/eslint/base.js
Comment (harishv7) [inline tooling/eslint/base.js:90]: why do we allow empty functions? do the tests require this?

File: apps/studio/src/utils/ajv.ts
Comment (harishv7) [inline apps/studio/src/utils/ajv.ts:3]: maybe not related to this PR, why do we use strict false?

File: apps/studio/src/server/modules/resource/resource.service.ts
Comment (harishv7) [inline apps/studio/src/server/modules/resource/resource.service.ts:143]: why do we not opt to throw here? wouldn't this be a 404?

File: apps/studio/tests/integration/helpers/seed/index.ts
Comment (harishv7) [inline apps/studio/tests/integration/helpers/seed/index.ts:9]: these 3 ops seems to be fetching the existing site, nav, footer configs. Why are they within the setupSite function?

File: apps/studio/tests/integration/helpers/seed/index.ts
Comment (harishv7) [inline apps/studio/tests/integration/helpers/seed/index.ts:134]: would it make sense to extract all of the values into constants to keep these queries clean (similar to curr seed script)

File: apps/studio/prisma/types.ts
Comment (harishv7) [inline apps/studio/prisma/types.ts:21]: Can share why using tagged/unwraptagged is better here?

File: apps/studio/src/schemas/page.ts
Comment (karrui) [reply apps/studio/src/schemas/page.ts:63]: it's ok, this will still cause the validation to fail. FATAL just means zod stops validating immediately. No big deal for it to gather all validation errors before returning to the user

File: apps/studio/src/server/modules/site/site.service.ts
Comment (karrui) [reply apps/studio/src/server/modules/site/site.service.ts:41]: It still worked because of luck. we did not trigger any ambiguity on the pg engine. In more complex shapes, postgres may throw. So this is insurance.

File: tooling/eslint/base.js
Comment (karrui) [reply tooling/eslint/base.js:90]: some tests, especially skipped tests, will trigger it. No need for lint to be super strict on tests too

File: apps/studio/src/utils/ajv.ts
Comment (karrui) [reply apps/studio/src/utils/ajv.ts:3]: cause there may be more metadata on the input than exactly the schema, which causes ajv to throw. this one need ask @dcshzj maybe is code smell.

File: apps/studio/src/server/modules/resource/resource.service.ts
Comment (karrui) [reply apps/studio/src/server/modules/resource/resource.service.ts:143]: I waffling between trpc errors being thrown inside `service` files or in the router files directly. Might want the caller to take responsibility instead (which the caller actually does in this case)

File: apps/studio/tests/integration/helpers/seed/index.ts
Comment (karrui) [reply apps/studio/tests/integration/helpers/seed/index.ts:9]: a site is always the triple <site, footer, navbar>. so if fetching we fetch lo; if not we create

File: apps/studio/tests/integration/helpers/seed/index.ts
Comment (karrui) [reply apps/studio/tests/integration/helpers/seed/index.ts:134]: next time ba. i think not unwieldy yet and still readable. I favor readability over everything for tests

File: apps/studio/prisma/types.ts
Comment (karrui) [reply apps/studio/prisma/types.ts:21]: Added an ADR 
https://www.notion.so/opengov/Explicitly-tagging-Kysely-jsonb-types-11577dbba78880c9917cd65dfe148e07?pvs=4

File: apps/studio/src/server/modules/page/__tests__/page.router.test.ts
Comment (seaerchin) [inline apps/studio/src/server/modules/page/__tests__/page.router.test.ts:23]: til -this is a really cool helper function!

File: apps/studio/src/server/modules/page/__tests__/page.router.test.ts
Comment (seaerchin) [inline apps/studio/src/server/modules/page/__tests__/page.router.test.ts:35]: actually this isn't quite right - we should also reset the `Resource` table to avoid inter-test pollution right? 

we don't want to have a later test being reliant on an earlier one

File: apps/studio/src/server/modules/page/__tests__/page.router.test.ts
Comment (seaerchin) [inline apps/studio/src/server/modules/page/__tests__/page.router.test.ts:250]: sorry for burden  - this test case isn't required for this pr but if you decide to not do it, could you file an issue and assign it to me?

we should check the behaviour in the event of value equality but different ordering of properties. as an example, consider the state `[{a:1, b:2}, some_other_block]`. if this gets changed to `[{b:2, a:1}, some_other_block]`, should we consider the state as being changed? to the user, this will appear exactly the same (same properties on block and block is same index) but i think this will fail atm cos we do a string comparison

**Review by seaerchin — DISMISSED** [review]
lgtm! actually 1 thing that's risky right now but untested is the behaviour of js coercion to pg queries. 

an example we have atm, is that we coerce `Number(some_str)`, which might result in `NaN`. the converse, `String(some_number)` is safer but do you think we should add in a test case for the former or even both?

**Comment by seaerchin** [comment]
eh approve too early omegalul

File: apps/studio/src/server/modules/page/__tests__/page.router.test.ts
Comment (seaerchin) [inline apps/studio/src/server/modules/page/__tests__/page.router.test.ts:543]: @harishv7 trouble you to relook again at the draft logic. we should aim to have our invariants verified through test

File: apps/studio/src/server/modules/page/__tests__/page.router.test.ts
Comment (seaerchin) [inline apps/studio/src/server/modules/page/__tests__/page.router.test.ts:573]: leave it to you but shuold we check if this id is different from the published blob id? this is to check that we don't override already published blobs

File: apps/studio/src/server/modules/page/__tests__/page.router.test.ts
Comment (seaerchin) [inline apps/studio/src/server/modules/page/__tests__/page.router.test.ts:790]: actually our invariant is that pages can only be created if your parent is `null` or a folder, so we need to check that `folderId` is not a collection too

File: apps/studio/src/server/modules/page/__tests__/page.router.test.ts
Comment (seaerchin) [inline apps/studio/src/server/modules/page/__tests__/page.router.test.ts:890]: @harishv7 - i think it'd be good for us to write **test descriptors**, wdyt? can you write skipped tests for behaviour that's important to verify here?

File: apps/studio/src/server/modules/page/__tests__/page.router.test.ts
Comment (seaerchin) [inline apps/studio/src/server/modules/page/__tests__/page.router.test.ts:822]: low value case but should we check behaviour of what happens when the `parentId` of rootpage is non-null?

technically this should return a 5xx but i think it'll pass at present. although it's worth noting that actually users don't get to edit this property so it's really really really very minor and i'm happy to skip

File: apps/studio/src/server/modules/page/__tests__/page.router.test.ts
Comment (seaerchin) [inline apps/studio/src/server/modules/page/__tests__/page.router.test.ts:892]: should we also check that permalink cannot have special characters?

File: apps/studio/src/server/modules/page/page.router.ts
Comment (seaerchin) [inline apps/studio/src/server/modules/page/page.router.ts:246]: wdyt about "The homepage for your site could not be found"? `rootpage` is internal to isomer and we should keep terminology consistent cc @sehyunidaaa

File: apps/studio/src/server/modules/page/page.router.ts
Comment (seaerchin) [inline apps/studio/src/server/modules/page/page.router.ts:377]: sadly we should revert this because this error is likely going to be seen by the user and the place where it shows is in `/settings` for the permalink .-.

File: apps/studio/src/server/modules/resource/resource.service.ts
Comment (seaerchin) [inline apps/studio/src/server/modules/resource/resource.service.ts:191]: no real opinion here - noticed previous call didn't throw error but this now does. not sure how muych value consistency brings here but if nextjs has good error handling layer, i think it's ok to throw in services

File: apps/studio/src/server/modules/resource/resource.service.ts
Comment (seaerchin) [inline apps/studio/src/server/modules/resource/resource.service.ts:388]: how would we distinguish between a resource being empty and a resource not existing in this case?

File: apps/studio/src/server/modules/resource/resource.service.ts
Comment (seaerchin) [inline apps/studio/src/server/modules/resource/resource.service.ts:422]: just a sanity check to ensure that this `null` value is handled properly on frontend

File: apps/studio/src/utils/ajv.ts
Comment (seaerchin) [reply apps/studio/src/utils/ajv.ts:3]: i think this is because we need to have extra properties on the item that ajv doesn't support. i could be wrong here thoguh

**Review by seaerchin — APPROVED** [review]
lgtm! i think making progress here is better than making no progress - for the test cases that i've raised, let me know if you are going to be doing it here or file issue. both are fine but we should just track them somewhere if it's hindering progress

**Comment by seaerchin** [comment]
actually just fix the copy before merge is ok already

File: apps/studio/src/server/modules/page/__tests__/page.router.test.ts
Comment (seaerchin) [reply apps/studio/src/server/modules/page/__tests__/page.router.test.ts:890]: https://github.com/opengovsg/isomer/pull/770

**Comment by karrui** [comment]
### Merge activity

* **Oct 11, 5:55 AM EDT**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/opengovsg/isomer/720).

**Pull request merged**
